components/nexus-cellar-hazelcast/src/main/java/org/sonatype/nexus/hazelcast/internal/HazelcastInstanceProvider.java

```diff
@@ -0,0 +1,171 @@
+import java.util.Map;
+
+import javax.annotation.PreDestroy;
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Provider;
+
+import org.sonatype.nexus.common.app.ApplicationDirectories;
+import org.sonatype.nexus.hazelcast.HazelcastMemberAttributesCustomizer;
+import org.sonatype.sisu.goodies.common.ComponentSupport;
+
+import com.google.common.base.Throwables;
+import com.hazelcast.config.Config;
+import com.hazelcast.config.MemberAttributeConfig;
+import com.hazelcast.config.XmlConfigBuilder;
+import com.hazelcast.core.Hazelcast;
+import com.hazelcast.core.HazelcastInstance;
+import com.hazelcast.util.ExceptionUtil;
+import org.apache.karaf.cellar.core.utils.CombinedClassLoader;
+import org.eclipse.sisu.EagerSingleton;
+import org.osgi.framework.BundleContext;
+import org.osgi.framework.ServiceRegistration;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+/**
+ * Expose the shared {@link HazelcastInstance}.
+ *
+ * @since 3.0
+ */
+@Named(HazelcastInstanceProvider.INSTANCE_NAME)
+@EagerSingleton
+public class HazelcastInstanceProvider
+    extends ComponentSupport
+    implements Provider<HazelcastInstance>
+{
+  public static final String INSTANCE_NAME = "nexus";
+
+  static {
+    // HACK: workaround https://github.com/orientechnologies/orientdb/pull/4609
+    ExceptionUtil.class.toString();
+  }
+
+  private final BundleContext bundleContext;
+
+  private final ApplicationDirectories directories;
+
+  private final List<HazelcastMemberAttributesCustomizer> memberAttributesCustomizers;
+
+  private final CombinedClassLoader combinedClassLoader;
+
+  private HazelcastInstance instance;
+
+  private ServiceRegistration<?> registration;
+
+  @Inject
+  public HazelcastInstanceProvider(final BundleContext bundleContext,
+                                   final ApplicationDirectories directories,
+                                   final List<HazelcastMemberAttributesCustomizer> memberAttributesCustomizers)
+  {
+    this.bundleContext = checkNotNull(bundleContext);
+    this.directories = checkNotNull(directories);
+    this.memberAttributesCustomizers = checkNotNull(memberAttributesCustomizers);
+    this.combinedClassLoader = new CombinedClassLoader();
+
+    // start tracking bundles whose packages should be distributed
+    new HazelcastBundleTracker(bundleContext, combinedClassLoader);
+  }
+
+  @Override
+  public HazelcastInstance get() {
+    instance = Hazelcast.newHazelcastInstance(getHazelcastConfig());
+    log.info("Instance: {}", instance);
+
+    // expose hazelcast instance as service so cellar's blueprint components can consume it
+    registration = bundleContext.registerService(HazelcastInstance.class, instance, null);
+
+    return instance;
+  }
+
+  @PreDestroy
+  public void shutdown() {
+    if (instance != null) {
+      log.info("Shutting down: {}", instance);
+      try {
+        registration.unregister();
+      }
+      catch (final IllegalStateException e) {
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) Either log or rethrow this exception. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1166)